### PR TITLE
Remove memory copy in diag function

### PR DESCRIPTION
**Pull request by okuta**
This PR is blocked by #1124.

### CODE REVIEW

File: cupy/creation/matrix.py [hunk 1]
```diff
@@ -22,24 +22,25 @@ def diag(v, k=0):
...
     if ndim == 1:
         size = v.size + abs(k)
         ret = cupy.zeros((size, size), dtype=v.dtype)
         ret.diagonal(k)[:] = v
         return ret
-    else:
+    elif ndim == 2:
         return cupy.array(v.diagonal(k))
+    else:
+        raise ValueError("Input must be 1- or 2-d.")
```
Comment (unnonouno) [inline cupy/creation/matrix.py]: Add test with ndim=3.

File: tests/cupy_tests/creation_tests/test_matrix.py [hunk 2]
```diff
@@ -56,6 +56,10 @@ def test_diag_construction_from_tuple(self, xp):
         self.assertIsInstance(r, xp.ndarray)
         return r
 
+    @testing.numpy_cupy_raises()
+    def test_diag_scaler(self, xp):
+        return xp.diag(1)
```
Comment (unnonouno) [inline tests/cupy_tests/creation_tests/test_matrix.py]: Add test with zero-dim cupy array and zero-dim numpy array.

**Comment by okuta** [comment]
I added tests

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**